test/test_bundle_importer.js

```diff
@@ -191,26 +213,70 @@ describe('Bundle Importer', function() {
         addUnit: true,
-        addRelation: true
+        addRelation: true,
+        setAnnotations: true
       });
     });
-
-    it('does not fail if unknown record type', function() {
+    it('properly sorts a recordSet', function() {
+      var data = utils.loadFixture(
+          'data/wordpress-bundle-recordset.json', true);
+      var sorted = bundleImporter._sortDryRunRecords(data);
+      var order = [
+        'addCharm-0', 'addService-1', 'setAnnotations-2',
+        'addCharm-3', 'addService-4', 'setAnnotations-5',
+        'addCharm-6', 'addService-7', 'setAnnotations-8',
+        'addMachines-9', 'addMachines-10',
+        'addRelation-11', 'addRelation-12',
+        'addUnit-13', 'addMachines-16',
+        'addUnit-14', 'addMachines-17',
+        'addUnit-15'
+      ];
+      sorted.forEach(function(record, index) {
+        assert.equal(record.id, order[index]);
+      });
+    });
+    it('stops but does not fail if unknown record type', function() {
       var sortedRecords = [
         { method: 'badMethod' }
       ];
+      var execute = utils.makeStubMethod(bundleImporter, '_executeRecord');
+      var notification = utils.makeStubMethod(
+          bundleImporter.db.notifications, 'add');
+      this._cleanups.concat([execute.reset, notification.reset]);
       bundleImporter._executeDryRun(sortedRecords);
+      execute.passThroughToOriginalMethod();
+      // the executor should only be called once at which time it'll throw a
+      // notification instead of continuing on.
+      assert.equal(execute.callCount(), 1, 'execute not called');
+      assert.equal(notification.callCount(), 1, 'notification not added');
     });
+  });
 
-    describe('_execute_addCharm', function() {
-      it('loads a charm');
-      it('shows notification on failure');
+  describe('Changeset execution', function() {
+    it('Sets up the correct environment (Integration)', function() {
+      var data = utils.loadFixture(
+          'data/wordpress-bundle-recordset.json', true);
+      bundleImporter.importBundleDryRun(data);
+      assert.equal(db.services.size(), 3);
+      assert.equal(db.machines.size(), 4);
+      assert.equal(db.relations.size(), 2);
+      // Services
+      assert.equal(db.services.item(0).get('charm'), 'cs:precise/haproxy-35');
+      assert.equal(db.services.item(1).get('charm'), 'cs:precise/wordpress-27');
+      assert.equal(db.services.item(2).get('charm'), 'cs:precise/mysql-51');
+      // Machines
+      assert.equal(db.machines.item(0).id, 'new0');
+      assert.equal(db.machines.item(1).id, 'new1');
+      assert.equal(db.machines.item(2).id, 'new0/lxc/new3');
+      assert.equal(db.machines.item(3).id, 'new1/lxc/new2');
+      // Relations
+      assert.equal(
+          db.relations.item(0).get('id'),
+          'pending-$addService-1:reverseproxy$addService-4:website');
+      assert.equal(
+          db.relations.item(1).get('id'),
+          'pending-$addService-4:db$addService-7:db');
```

Comment:
Can you do an XXX here then with the known issue and then remove it in your follow up branch?